doc/manual/src/architecture/architecture.md

```diff
@@ -0,0 +1,42 @@
+# Architecture
+
+This chapter describes how Nix works.
+It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.
+
+## Overview
+
+Nix consists of hierarchical [layers](https://en.m.wikipedia.org/wiki/Multitier_architecture#Layers).
+
+```
+                   [ commmand line interface ]--------+
+                                |                     |
+                                | evaluates           | manages
+                                V                     |
+                   [ configuration language  ]        |
+                                |                     V
++-------------------------------|---------------------------------+
+| store                         |                                 |
+|                               | evaluates to                    |
+|                               |                                 |
+|               referenced by   V       builds                    |
+|   [ build input ] ----> [ build plan ] ----> [ build result ]   |
+|                                                                 |
++-----------------------------------------------------------------+
+```
+
+At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.
+
+Below that is the [Nix expression language](../expressions/expression-language.md), a [purely functional](https://en.m.wikipedia.org/wiki/Purely_functional_programming) configuration language.
+It is used to compose expressions which ultimately evaluate to self-contained *build plans*, used to derive *build results* from referenced *build inputs*.
+
+::: {.note}
+The Nix language itself does not have a notion of *packages* or *configurations*.
```

Comment:
I've drafted a definition for *package* in https://github.com/NixOS/nix/issues/6507. It is a bit more minimal and more general than the de facto handling of "packages" by the Nix tools.
Nonetheless, it is not part of the language per se, as it is a convention or interface around expressions.